lib/client.js

```diff
@@ -305,7 +306,14 @@ Client.prototype.resume = function () {
 
 function enqueue (packet) {
   this.client._parsingBatch++
-  handle(this.client, packet, this.client._nextBatch)
+  // already connected or it's the first packet
+  if (this.client.connackSent || this.client._parsingBatch === 1) {
```

Comment:
Is it possible to use the `connackSent` event here instead of using a new event in `connected`?

Reply:
We can't. `'connackSent'`  is delivered too early, and due to the now-fixed asynchronicity of Readable, some packets might arrive early on. We need to wait for the full auth to complete, and that is async.

Note that this was *needed* also with previous versions, and it might be the cause of some obscure bugs/security vulnerabilities.